sql/core/src/test/scala/org/apache/spark/sql/sources/v2/SimpleWritableDataSource.scala

```diff
@@ -0,0 +1,254 @@
+
+    override def createInternalRowWriterFactory(): DataWriterFactory[InternalRow] = {
+      new InternalRowCSVDataWriterFactory(path, jobId, new SerializableConfiguration(conf))
+    }
+
+    override def commit(messages: Array[WriterCommitMessage]): Unit = {
+      val finalPath = new Path(path)
+      val jobPath = new Path(new Path(finalPath, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      try {
+        for (file <- fs.listStatus(jobPath).map(_.getPath)) {
+          fs.rename(file, new Path(finalPath, file.getName))
+        }
+      } finally {
+        fs.delete(jobPath, true)
+      }
+    }
+
+    override def abort(messages: Array[WriterCommitMessage]): Unit = {
+      val jobPath = new Path(new Path(path, "_temporary"), jobId)
+      val fs = jobPath.getFileSystem(conf)
+      fs.delete(jobPath, true)
+    }
+  }
+
+  override def createReader(options: DataSourceV2Options): DataSourceV2Reader = {
+    val path = new Path(options.get("path").get())
+    val conf = SparkContext.getActive.get.hadoopConfiguration
+    new Reader(path.toUri.toString, conf)
+  }
+
+  override def createWriter(
+      schema: StructType,
+      mode: SaveMode,
+      options: DataSourceV2Options): Optional[DataSourceV2Writer] = {
+    assert(DataType.equalsStructurally(schema.asNullable, this.schema.asNullable))
+    assert(!SparkContext.getActive.get.conf.getBoolean("spark.speculation", false))
+
+    val path = new Path(options.get("path").get())
+    val internal = options.get("internal").isPresent
+    val conf = SparkContext.getActive.get.hadoopConfiguration
+    val fs = path.getFileSystem(conf)
+    if (fs.exists(path)) {
+      if (mode == SaveMode.ErrorIfExists) {
+        throw new RuntimeException("data already exists.")
+      }
+
+      if (mode == SaveMode.Ignore) {
+        Optional.empty()
+      } else if (mode == SaveMode.Overwrite) {
+        fs.delete(path, true)
+        Optional.of(createWriter(path, conf, internal))
+      } else {
+        assert(mode == SaveMode.Append)
+        Optional.of(createWriter(path, conf, internal))
+      }
+    } else {
+      Optional.of(createWriter(path, conf, internal))
+    }
+  }
+
+  private def createWriter(
+      path: Path, conf: Configuration, internal: Boolean): DataSourceV2Writer = {
+    val pathStr = path.toUri.toString
+    if (internal) new InternalRowWriter(pathStr, conf) else new Writer(pathStr, conf)
+  }
+}
+
+class SimpleCSVReadTask(path: String, conf: SerializableConfiguration)
+  extends ReadTask[Row] with DataReader[Row] {
+
+  @transient private var lines: Iterator[String] = _
+  @transient private var currentLine: String = _
+
+  override def createReader(): DataReader[Row] = {
+    val filePath = new Path(path)
+    val fs = filePath.getFileSystem(conf.value)
+    val in = fs.open(filePath)
+    lines = new BufferedReader(new InputStreamReader(in))
+      .lines().iterator().asScala.toArray.toIterator
+    in.close()
```

Comment:
Unless the input stream is fully read at this point, this closes the input stream and attempts to read lines() will fail. Better: delete this line and rely on some downstream close operation at the end of the iteration